src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.inline.hpp

```diff
@@ -419,9 +400,9 @@ ShenandoahScanRemembered<RememberedSet>::verify_registration(HeapWord* address,
     //        cannot use card_index_for_addr(base_addr + offset) because it asserts arg < end of whole heap
     size_t end_card_index = index + offset / CardTable::card_size_in_words();
 
-    if (end_card_index > index) {
+    if (end_card_index > index && end_card_index < _rs->last_valid_index()) {
```

Comment:
If _rs->last_valid_index() is actually a valid index, why couldn't we make this test <=?  Or should we rename the last_valid_index() to be index_beyond_end()?

Reply:
Good catch. I will test that.